dockside/io.py

```diff
@@ -1,53 +1,165 @@
 from pathlib import Path
-from six.moves.urllib.request import urlopen
 
+import requests
 import pandas as pd
 
-def _make_params(paramlist):
 
-    if not isinstance(paramlist, list):
-        paramlist = [paramlist]
-    param_dict = {
-        'flow': 'cb_00060=on&',
-        'temp': 'cb_00010=on&',
-        'stage': 'cb_00065=on&',
-        'precip': 'cb_00045=on&'
+def fetch_nwis(site, start, end, daily=False, **kwargs):
+    """ Fetch JSON data from NWIS
+
+    Parameters
+    ----------
+    site : int, string, or sequence
+        Site ID number from NWIS. E.g, site = 14211500 or site = '14211500' for
+        Johnson Creek in Portland, OR. This can also be a list-like object for
+        multiple sites (experimental).
+    start, end : string or date-like
+        Some form of date representation for the start and end of the NWIS
+        observations you'd like to download
+    daily : bool (default is False)
+        Toggles downloading daily (True) or instanteous values (False, default)
+
+    Additional Parameters
+    ---------------------
+    All additional keyword arguments are passed directly to the NWIS API.
+
+    Returns
+    -------
+    request
+        A request object representing the API call to fetch the data.
+
+    Examples
+    --------
+    >>> import json
+    >>> from tempfile import TemporaryDirectory
+    >>> from pathlib import Path
+    >>> from dockside.io import fetch_nwis, read_nwis
+    >>> r = fetch_nwis(14211500, '2018-01-01', '2018-06-30', daily=False)
+    >>> with TemporaryDirectory() as td:
+    ...     fpath = Path(td) / 'JCreek_flow.json'
+    ...     with fpath.open('w') as fp:
+    ...         json.dump(r.json(), fp)
+
+    """
+
+    dtfmt = '%Y-%m-%d'
+    url_base = "https://nwis.waterservices.usgs.gov/nwis/{}".format('dv' if daily else 'iv')
+    url_params = {
+        "format": kwargs.pop('format', 'json'),
+        "sites": site,
+        "startDT": pd.Timestamp(start).strftime(dtfmt),
+        "endDT": pd.Timestamp(end).strftime(dtfmt),
+        **kwargs
+    }
+
+    return requests.get(url_base, params=url_params)
+
+
+def _expand_columns(df, names, sep='_'):
+    """
+    Splits string column labels into tuples
+    """
+
+    newcols = df.columns.str.split(sep, expand=True)
+    return (
+        df.set_axis(newcols, axis='columns', inplace=False)
+          .rename_axis(names, axis='columns')
+    )
+
+
+def _parse_ts(ts, daily):
+    """
+    Parses a single `timeSeries` object in an NWIS JSON response in a dataframe
+    """
+
+    param = ts['variable']['variableName']
+    if daily:
+        stat = ts['variable']['options']['option'][0]['value']
```

Comment:
Previously there was a way to get all of the different station variables from the text scraping (e.g.:
        'flow': 'cb_00060=on&',
        'temp': 'cb_00010=on&',
        'stage': 'cb_00065=on&',
        'precip': 'cb_00045=on&')
Does the JSON return everything available?

Reply:
JSON returns everything that's available. you can use `**kwargs` to limit it